Add leak fuel from fueltanks event

GoDne0 opened the pull request:
Добавил ивент для сторителера с протечкой топливных баков.
Может протечь от 2 до 5 баков, баки остаются открытыми и при их перемещении так же будет вытекать топливо.
Для старта ивента требуется наличие хотя бы одного инженегра, проверка вроде даже работает

Не очень разобрался с тем, как работает get_mtth, так что буду рад помощи с цифорками, сейчас шанс ивента слишком большой.

close #10728

<details>
<summary>Чейнджлог</summary>

```yml
🆑GoDneo
rscadd: Добавлен ивент для сторителера с утечкой топлива из топливных баков.
/🆑
```

</details>

- [] Pull Request полностью завершен, мне не нужна помощь чтобы его закончить.
- [x] Я внимательно прочитал(-а) все свои изменения и багов в них не нашёл(-ла).
- [x] Я запускал(-а) сервер со своими изменениями локально и все протестировал(-а).
- [x] Я ознакомился(-ась) c [Guide to Contribute](https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md).

GoDne0 committed:
Leaking fufeltank

GoDne0 committed:
Pravki

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+
+	var/list/station_levels = GLOB.using_map.get_levels_with_trait(ZTRAIT_STATION)
+	var/list/matching_tanks = list()
+	announce()
```
Я не помню точно как работают ивенты, это точно не лишний код?

GoDne0 replied:
Если ты про announce(), то нет. Без него аннонс не работал

Filatelele commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -115,8 +115,26 @@
+/obj/structure/reagent_dispensers/fueltank/Move(newloc, direct)
+	. = ..()
+	var/turf/old_turf = get_turf(loc)
+	var/turf/new_turf = get_turf(newloc)
```
Очередной пример почему не надо тянуть всякую бяку к себе и править код инструментами, которые для этого не предназначены.
Здесь нам важно получить текущий лок до движения - и рассчитанный следующий лок. Это мы делаем до ``. = ..()`` то бишь вызова родителя. Если непонятно че делают эти ебобомные скобочки да точки - велкам в ЛС.
```suggestion
/obj/structure/reagent_dispensers/fueltank/Move(newloc, direct)
	var/turf/old_turf = get_turf(loc)
	var/turf/new_turf = get_turf(newloc)
	. = ..()
```

Filatelele commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -115,8 +115,26 @@
+
+	if(old_turf?.z != new_turf?.z)
+		GLOB.fueltanks[old_turf.z] -= src
+		GLOB.fueltanks[new_turf.z] += src
```
```suggestion
	if(old_turf?.z != new_turf?.z)
		LAZYREMOVEASSOC(GLOB.fueltanks, old_turf?.z, src)
		LAZYADDASSOC(GLOB.fueltanks, new_turf?.z, src)
```

Filatelele commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -115,8 +115,26 @@
 /obj/structure/reagent_dispensers/fueltank/Destroy()
 	QDEL_NULL(rig)
+	if(GLOB.fueltanks[z])
+		GLOB.fueltanks[z] -= src
```
```suggestion
	LAZYREMOVEASSOC(GLOB.fueltanks, Z, src)
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+			continue
+		matching_tanks += candidate
+
+	var/leaking_tanks_quantity = rand(2,5)
```
```suggestion
	var/leaking_tanks_quantity = rand(2, 5)
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+	for(var/obj/structure/reagent_dispensers/fueltank/candidate as anything in GLOB.fueltanks)
+		if(!(candidate.z in station_levels))
+			continue
+		matching_tanks += candidate
```
Сладкие ньюлайники 
```suggestion
			continue

		matching_tanks += candidate
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+	while(already_leaking_tanks < leaking_tanks_quantity)
+		if(!matching_tanks?.len)
+			break
+		var/obj/structure/reagent_dispensers/fueltank/fueltank_leaked = pick_n_take(matching_tanks)
```
Сладкие ньюлайники [2]
```suggestion
		if(!matching_tanks?.len)
			break

		var/obj/structure/reagent_dispensers/fueltank/fueltank_leaked = pick_n_take(matching_tanks)
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+		var/obj/structure/reagent_dispensers/fueltank/fueltank_leaked = pick_n_take(matching_tanks)
+		matching_tanks -= fueltank_leaked
+		if(fueltank_leaked)
+			var/fuel_units_leak = rand(25,100)
```
```suggestion
			var/fuel_units_leak = rand(25, 100)
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+
+	var/leaking_tanks_quantity = rand(2,5)
+	var/already_leaking_tanks = 0
+	while(already_leaking_tanks < leaking_tanks_quantity)
```
Тут уже от тебя зависит, мне слегка лень думать и это больше мой зуд в гузне чем конкретная претензия.

Я ненавижу while() ибо с их помощью значительно проще убивать билд по сравнению с for() 
Тут есть потенциал для очень долгой итерации по этому циклу по сравнению с тем решением которое я тебе в личку кидал.

Filatelele commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -115,8 +115,26 @@
 
+/obj/structure/reagent_dispensers/fueltank/Initialize()
+	. = ..()
+	LAZYADDASSOC(GLOB.fueltanks, src, z)
```
Я очень давно не трогал наш билд и уже, честно говоря, не помню синтаксис LAZYADDASSOC
Но, кажется, там аргументы подаются формата (list, key, value)
А значит - нам нужно сделать так. 
```suggestion
	LAZYADDASSOC(GLOB.fueltanks, z, src)
```

Но я мог обмануть, поэтому чекни как лейзиэдд работает сам.

intercepti0n replied:
`LAZYADDASSOC(L, K, V)`

Filatelele reviewed (changes requested):
В целом збс, оставил пару комментов и саггов.

GoDne0 committed:
Update code/modules/reagents/reagent_dispenser.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/modules/reagents/reagent_dispenser.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/modules/reagents/reagent_dispenser.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/datums/events/leaking_tanks.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/datums/events/leaking_tanks.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/datums/events/leaking_tanks.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update code/datums/events/leaking_tanks.dm

Co-authored-by: Filatelele <dfilatoff2014@yandex.ru>

GoDne0 committed:
Update leaking_tanks.dm

GoDne0 committed:
Update reagent_dispenser.dm

GoDne0 commented:
Заменил цикл while на for, подправил LAZYREMOVEASSOC и LAZYADDSOC, вроде всё работает. Нужно только с шансом на прок ивента разобрался, тут не понятно мне как оно считает

GoDne0 committed:
Update leaking_tanks.dm

GoDne0 committed:
Update leaking_tanks.dm

Потрогал цифорки, теперь проценты на прок ивента не набегают с огромной скоростью

GoDne0 commented:
Потрогал цифры
	mtth = 1 HOURS
	difficulty = 10
Теперь процент прока ивента не набегают с огромной скоростью

i1yadobr commented:
Я не эксперт, но...

![356363636246](https://github.com/user-attachments/assets/ecea761c-40e7-44c0-821b-b23d4a103c12)

Оно точно должно быть так?

![image](https://github.com/user-attachments/assets/d006b0fe-6dfb-45de-9eb0-673c17cc1044)

![474674747](https://github.com/user-attachments/assets/436e033f-b9f4-4ea0-a6b8-c74379b3bec3)

Я если что не пытался разобраться в коде, но одним глазком подметил такую деталь. Если всё правильно работает, можно это сообщение игнорировать.

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+
+/datum/event/leaking_tanks/get_mtth()
+	. = ..()
+	. -= (SSevents.triggers.roles_count["Station Engineer"] * (15 MINUTES))
```
```suggestion
	. -= (SSevents.triggers.roles_count["Engineer"] * (15 MINUTES))
```

Filatelele commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+	. = max(1 HOUR, .)
+
+/datum/event/leaking_tanks/check_conditions()
+	. = SSevents.triggers.roles_count["Station Engineer"] >= 1
```
```suggestion
	. = SSevents.triggers.roles_count["Engineer"] >= 1
```

Filatelele commented:
Наверное Илья прав, хз. Надо разбираться, но так впадлу - будем считать, что он действительно прав.

Filatelele committed:
Update code/datums/events/leaking_tanks.dm

Filatelele committed:
Update code/datums/events/leaking_tanks.dm

Filatelele reviewed (dismissed):
![image](https://github.com/user-attachments/assets/16e536af-e199-4623-b4ac-432ca22c099c)

GoDne0 commented:
Я бы оставил Station Engineer, проверил на локалке доступность ивента раундстартом с ["Engineer"]:
![1](https://github.com/user-attachments/assets/ea05e01d-71ee-4f4f-9689-b8ad8080994b)
и c ["Station Engineer"]:
![2](https://github.com/user-attachments/assets/371cb464-a643-4ba3-8917-6a7d275cf11a)

i1yadobr commented:
Тебе бы следовало поменять это значение _на прежнее_ перед тем как залить изменения, наверное.
Но тут тоже не уверен, если честно. Тебе уже поставили галочку, так что решай сам - оставить, или забить.

![image](https://github.com/user-attachments/assets/1f022cee-63b7-4a67-a7e1-c321fbac6021)

GoDne0 committed:
Update admin.toml

GoDne0 committed:
Update leaking_tanks.dm

GoDne0 commented:
Спасибо Илья, забыл про это, поправил.
Вернул пока Station Engineer на место ибо без него не запускается ивент при наличии раундстарового инженегра. Наверное, в других ивентах должна идти проверка на наличие или шанс срабатывания от кол-ва инженеров, с них код и брал, но я не уверен что там оно вообще работает. Думаю тут лучше оставить так пока работает.

i1yadobr commented:
@GoDne0 
И на будущее. Чтобы тебе не приходилось миллион раз менять эту переменную туда-сюда, сделай проще:

![dev](https://github.com/user-attachments/assets/4cb1f01b-8510-4728-982e-ffc8e2ee5b0f)

Залезь в свойства системы, переменные среды, создай переменную "MODE", и туда значение "dev".
Так у тебя конфиг будет подтягивать значение, которые у тебя находятся по пути config/dev, где, собственно, это значение уже выставлено:

![3563573573](https://github.com/user-attachments/assets/fd73852b-7d58-4468-9e43-d317eb8d8a75)

Это также касается новых изменений, если они тебе когда-нибудь понадобятся.
Для более подробной информации можешь сюда перекатиться:
https://github.com/ChaoticOnyx/OnyxBay/blob/dev/docs/contributing.md#конфигурация
Также ты мо-о-о-ожешь, конечно, поднять собственную базу данных, но думаю тебе оно не понадобиться, если ты не планируешь работать в сфере SQL.
Также не уверен, какая у тебя там система, но если что поищи что-то связанное с "переменная среда" и сделай как тебе удобнее.
Я чисто тебе пример со своей системы показал, как оно у меня.

GoDne0 committed:
Update leaking_tanks.dm

GoDne0 committed:
Update announce.dm

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+	mtth = 1 HOURS
+	difficulty = 10
+
+/datum/event/leaking_tanks/New()
```
Убери, в этом коде нет смысла.

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,48 @@
+/datum/event/leaking_tanks/on_fire()
+	var/list/station_levels = GLOB.using_map.get_levels_with_trait(ZTRAIT_STATION)
+	var/list/matching_tanks = list()
+	var/leaking_tanks_quantity = rand(2, 5)
```
Перенеси объявление ближе к циклу, в котором она используется. Мы не на Си пишем все-таки.

intercepti0n commented on code/modules/reagents/reagent_dispenser.dm:
```diff
@@ -115,8 +115,25 @@
+/obj/structure/reagent_dispensers/fueltank/Initialize()
+	. = ..()
+	LAZYADDASSOC(GLOB.fueltanks, src, z)
+	return
```
Нет смысла писать `return`, если далее в функции ничего не происходит.

intercepti0n commented on code/_global_vars/lists/objects.dm:
```diff
@@ -28,3 +28,4 @@ GLOBAL_DATUM_INIT(sun, /datum/sun, new)
 GLOBAL_DATUM_INIT(universe, /datum/universal_state, new)
 
 GLOBAL_LIST_EMPTY(intact_station_closets) // List of closets (excluding crates) located on the STATION Z-LEVELS that have never been opened since initialization. Mostly for the Gatecrasher event, but may be useful for more fun thingies.
+GLOBAL_LIST_EMPTY(fueltanks)	//list of fueltanks on station z-levels. Mostly need for leaking fueltank event
```
Перенеси на строку выше коммент и добавь третий слэш, чтобы к документация цеплялась расширением.

GoDne0 committed:
Update leaking_tanks.dm

GoDne0 committed:
Update objects.dm

GoDne0 committed:
Update reagent_dispenser.dm

GoDne0 committed:
some code fix

GoDne0 committed:
another code fix

GoDne0 committed:
another code fix 2

GoDne0 committed:
another code fix 3

GoDne0 committed:
another code fix 4

Надеюсь последний комит

GoDne0 committed:
another code fix 5

Я заебался исправлять по паре символов за коммит, но оно работает

GoDne0 committed:
final code fix maybe

GoDne0 committed:
И ещё правки

GoDne0 committed:
...

GoDne0 committed:
another commit

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,49 @@
+	announce()
+
+	for (var/i in 1 to rand(MIN_TANK_QUANTITY, MAX_TANK_QUANTITY))
+
```
```suggestion
```

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,49 @@
+	. = max(1 HOUR, .)
+
+/datum/event/leaking_tanks/on_fire()
+
```
```suggestion
```

intercepti0n commented on code/datums/events/leaking_tanks.dm:
```diff
@@ -0,0 +1,49 @@
+	for (var/z_level as anything in GLOB.using_map.get_levels_with_trait(ZTRAIT_STATION))
+		LAZYADD(matching_tanks, LAZYCOPY(GLOB.fueltanks["[z_level]"]))
+
+	if(!(length(matching_tanks)))
```
```suggestion
	if (!length(matching_tanks))
```

intercepti0n committed:
Update code/datums/events/leaking_tanks.dm

intercepti0n committed:
Update code/datums/events/leaking_tanks.dm

intercepti0n committed:
Update code/datums/events/leaking_tanks.dm
